storage: wait on lease renewal when in stasis, kv: retry not leaseholder errors

spencerkimball opened the pull request:
In storage, discovered a bug where we would fail to wait on a lease renewal
for a range where the existing lease was in state=stasis.

In kv, `DistSender.sendToReplicas` now retries replicas which previously
failed with retryable errors (right now this is just `NotLeaseHolderErrors`).

Fixes #11221

spencerkimball committed:
storage: wait on lease renewal when in stasis, kv: retry not leaseholder errors

In storage, discovered a bug where we would fail to wait on a lease renewal
for a range where the existing lease was in state=stasis.

In kv, `DistSender.sendToReplicas` now retries replicas which previously
failed with retryable errors (right now this is just `NotLeaseHolderErrors`).

Fixes #11221

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/15199)
<!-- Reviewable:end -->

petermattis commented:
Review status: 0 of 7 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[pkg/kv/transport.go, line 158 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15199#-KiAkBgeg4Ke75QF0XwA:-KiAkBgeg4Ke75QF0XwB:b-iilog) ([raw file](https://github.com/cockroachdb/cockroach/blob/e83e89348f7817513ab1fe7e4a9ba8597f784407/pkg/kv/transport.go#L158)):*
> ```Go
> }
> 
> // IsExhausted first attempts to resurrect any replicas which were
> ```

Well, it checks to see if there are any untried replicas first.

---

*[pkg/kv/transport.go, line 197 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15199#-KiAkefYoPwq254WUJFJ:-KiAkefYoPwq254WUJFK:bycxhf0) ([raw file](https://github.com/cockroachdb/cockroach/blob/e83e89348f7817513ab1fe7e4a9ba8597f784407/pkg/kv/transport.go#L197)):*
> ```Go
> 		return defaultTimeout, true
> 	}
> 	var timeout time.Time
> ```

Mildly confusing that this `timeout` is a `time.Time` while other timeouts are `time.Duration`. Might want to rename this `deadline` as well as the `batchClient.timeout` field.

---

*[pkg/kv/transport.go, line 200 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15199#-KiAkXJzi4B9WoCozJM_:-KiAkXJzi4B9WoCozJMa:b-4tl2us) ([raw file](https://github.com/cockroachdb/cockroach/blob/e83e89348f7817513ab1fe7e4a9ba8597f784407/pkg/kv/transport.go#L200)):*
> ```Go
> 	for i := 0; i < gt.clientIndex; i++ {
> 		if c := gt.orderedClients[i]; !c.pending && c.retryable {
> 			if (timeout == time.Time{}) || c.timeout.Sub(timeout) < 0 {
> ```

I think this can be replaced by:

```
if timeout.Before(c.timeout) {
  timeout = c.timeout
}
```

---

*[pkg/kv/transport.go, line 208 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15199#-KiAlJ7O5UZkrxjFgr4D:-KiAlJ7O5UZkrxjFgr4E:b-1mfzay) ([raw file](https://github.com/cockroachdb/cockroach/blob/e83e89348f7817513ab1fe7e4a9ba8597f784407/pkg/kv/transport.go#L208)):*
> ```Go
> 		return 0, false
> 	}
> 	return timeout.Sub(timeutil.Now()), true
> ```

Is it problematic if this returns a negative value?

---

*[pkg/storage/replica.go, line 1045 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15199#-KiAjCSlYYaLYXExvikq:-KiAjCSlYYaLYXExvikr:badyj4) ([raw file](https://github.com/cockroachdb/cockroach/blob/e83e89348f7817513ab1fe7e4a9ba8597f784407/pkg/storage/replica.go#L1045)):*
> ```Go
> 				// ignore the returned channel (which is buffered) and continue.
> 				if status.state == leaseStasis {
> 					return r.requestLeaseLocked(ctx, status), nil
> ```

I didn't follow all of the code paths, but it seems strange to be requesting the lease here without first checking if there is a pending lease request. The previous location of this check seemed preferable, but clearly it was missing some condition and I'm not understanding what that condition is.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15199)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Review status: 0 of 7 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[pkg/kv/transport.go, line 158 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15199#-KiAkBgeg4Ke75QF0XwA:-KiAx1beQWVS-Rh_Hfh9:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e83e89348f7817513ab1fe7e4a9ba8597f784407/pkg/kv/transport.go#L158)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Well, it checks to see if there are any untried replicas first.
</blockquote></details>

Done.

---

*[pkg/kv/transport.go, line 197 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15199#-KiAkefYoPwq254WUJFJ:-KiAxHzCZTGfJmgESUeG:bt1b19p) ([raw file](https://github.com/cockroachdb/cockroach/blob/e83e89348f7817513ab1fe7e4a9ba8597f784407/pkg/kv/transport.go#L197)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Mildly confusing that this `timeout` is a `time.Time` while other timeouts are `time.Duration`. Might want to rename this `deadline` as well as the `batchClient.timeout` field.
</blockquote></details>

Renamed.

---

*[pkg/kv/transport.go, line 200 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15199#-KiAkXJzi4B9WoCozJM_:-KiAxXD77ssOEUD3Cz6S:b-sm4xys) ([raw file](https://github.com/cockroachdb/cockroach/blob/e83e89348f7817513ab1fe7e4a9ba8597f784407/pkg/kv/transport.go#L200)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I think this can be replaced by:

```
if timeout.Before(c.timeout) {
  timeout = c.timeout
}
```
</blockquote></details>

Not quite. Unfortunately we still need the `timeout == time.Time{}` because we're actually finding the earliest deadline, so we need `c.timeout.Before(timeout)`. I changed to using `Before()` though.

---

*[pkg/kv/transport.go, line 208 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15199#-KiAlJ7O5UZkrxjFgr4D:-KiAxlYkeGv4CxMv6hzW:bdsq8xw) ([raw file](https://github.com/cockroachdb/cockroach/blob/e83e89348f7817513ab1fe7e4a9ba8597f784407/pkg/kv/transport.go#L208)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Is it problematic if this returns a negative value?
</blockquote></details>

No, that just causes the timer to fire as though you set 0.

---

*[pkg/storage/replica.go, line 1045 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15199#-KiAjCSlYYaLYXExvikq:-KiAxt_l5EdBGmsJ8SkK:bu5tp3d) ([raw file](https://github.com/cockroachdb/cockroach/blob/e83e89348f7817513ab1fe7e4a9ba8597f784407/pkg/storage/replica.go#L1045)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I didn't follow all of the code paths, but it seems strange to be requesting the lease here without first checking if there is a pending lease request. The previous location of this check seemed preferable, but clearly it was missing some condition and I'm not understanding what that condition is.
</blockquote></details>

`requestLeaseLocked` will join the pending request and return the channel to wait on. Perhaps the naming of methods has become confusing here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15199)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: 0 of 12 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[pkg/kv/dist_sender_test.go, line 116 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/15199#-KiB0BkPJoltBCxY80qU:-KiB0BkPJoltBCxY80qV:b-j7hai8) ([raw file](https://github.com/cockroachdb/cockroach/blob/5832aedca608197cd9b6585c165e5fbeed223c51/pkg/kv/dist_sender_test.go#L116)):*
> ```Go
> }
> 
> func (l *legacyTransportAdapter) SendNextTimeout(defaultTimeout time.Duration) (time.Duration, bool) {
> ```

I don't get it; seems like none of the implementations ever return anything other than `defaultTimeout`, or false. I'd take the param out of the interface.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15199)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 12 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/base/constants.go, line 35 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/15199#-KiB1IEMLsZzGo8-HJdE:-KiB1IEMLsZzGo8-HJdF:b-k3z93z) ([raw file](https://github.com/cockroachdb/cockroach/blob/5832aedca608197cd9b6585c165e5fbeed223c51/pkg/base/constants.go#L35)):*
> ```Go
> 	DefaultHeartbeatInterval = 1 * time.Second
> 
> 	// DefaultSendNextTimeout is the amount wait duration before trying
> ```

`s/amount wait duration/amount of time to wait/g` 

---

*[pkg/storage/replica.go, line 1045 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15199#-KiAjCSlYYaLYXExvikq:-KiB1bmCpdF6TuLg6ay4:bmg03wb) ([raw file](https://github.com/cockroachdb/cockroach/blob/e83e89348f7817513ab1fe7e4a9ba8597f784407/pkg/storage/replica.go#L1045)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary><blockquote>

`requestLeaseLocked` will join the pending request and return the channel to wait on. Perhaps the naming of methods has become confusing here.
</blockquote></details>

Ah, so it does. Yes, the naming is unfortunate. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15199)*
<!-- Sent from Reviewable.io -->

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 12 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15199#-:-KiB1hyY7-k7AYy0cslb:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 6 of 7 files at r1, 8 of 8 files at r2.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[pkg/base/constants.go, line 35 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/15199#-KiB1IEMLsZzGo8-HJdE:-KiB1uWDwn5a5kPrb_02:b5wnatw) ([raw file](https://github.com/cockroachdb/cockroach/blob/5832aedca608197cd9b6585c165e5fbeed223c51/pkg/base/constants.go#L35)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

`s/amount wait duration/amount of time to wait/g` 
</blockquote></details>

s/amount of time/duration/ :)

---

*[pkg/kv/dist_sender.go, line 1207 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15199#-KiAwtADacl8V4VlVEq1:-KiAwtADacl8V4VlVEq2:brg4fbv) ([raw file](https://github.com/cockroachdb/cockroach/blob/8873c820a01214de1176e6c45d8ee48fec8deb46/pkg/kv/dist_sender.go#L1207)):*
> ```Go
> 						// If the replica we contacted knows the new lease holder, update the cache.
> 						ds.updateLeaseHolderCache(ctx, rangeID, *lh)
> ```

random whitespace change

---

*[pkg/kv/dist_sender_test.go, line 116 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/15199#-KiB0BkPJoltBCxY80qU:-KiB0moJlxye14jC9C8l:byha26i) ([raw file](https://github.com/cockroachdb/cockroach/blob/5832aedca608197cd9b6585c165e5fbeed223c51/pkg/kv/dist_sender_test.go#L116)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I don't get it; seems like none of the implementations ever return anything other than `defaultTimeout`, or false. I'd take the param out of the interface.
</blockquote></details>

These are test implementations. The real one is in kv/transport.

---

*[pkg/sql/ambiguous_commit_test.go, line 118 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15199#-KiAuB0ZRm5pR_0ZjE7t:-KiAuB0ZRm5pR_0ZjE7u:bgxqs29) ([raw file](https://github.com/cockroachdb/cockroach/blob/e83e89348f7817513ab1fe7e4a9ba8597f784407/pkg/sql/ambiguous_commit_test.go#L118)):*
> ```Go
> 				StoreID: tc.Servers[0].GetFirstStoreID(),
> 			})
> 		if err != nil {
> ```

unnecessary condition here, just return err

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15199)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 6 of 7 files at r1, 8 of 8 files at r2.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[pkg/kv/transport.go, line 163 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/15199#-KiB5d7dmWOiP_ynbKQ-:-KiB5d7dmWOiP_ynbKQ0:blhrjgg) ([raw file](https://github.com/cockroachdb/cockroach/blob/5832aedca608197cd9b6585c165e5fbeed223c51/pkg/kv/transport.go#L163)):*
> ```Go
> // which were tried but failed with a retryable error and have a
> // deadline set which has elapsed. If any where resurrected, returns
> // true; false otherwise.
> ```

The return values are backwards here.

---

*[pkg/kv/transport.go, line 326 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/15199#-KiB6IU5NnJ5UC6A9xEq:-KiB6IU5NnJ5UC6A9xEr:bnnybxy) ([raw file](https://github.com/cockroachdb/cockroach/blob/5832aedca608197cd9b6585c165e5fbeed223c51/pkg/kv/transport.go#L326)):*
> ```Go
> 			gt.orderedClients[i].retryable = retryable
> 			if retryable {
> 				gt.orderedClients[i].deadline = timeutil.Now().Add(gt.opts.SendNextTimeout)
> ```

Add a TODO to plumb the lease expiration through the NotLeaseHolderError so we can target the retry for after the old lease expires. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15199#-:-KiB73C5z4buHkxNGoa2:bnfp4nl)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[pkg/base/constants.go, line 35 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/15199#-KiB1IEMLsZzGo8-HJdE:-KiBEdHT98ZPQ8MKuXxT:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5832aedca608197cd9b6585c165e5fbeed223c51/pkg/base/constants.go#L35)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

s/amount of time/duration/ :)
</blockquote></details>

Done.

---

*[pkg/kv/dist_sender.go, line 1207 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15199#-KiAwtADacl8V4VlVEq1:-KiBEhOub4v9WfwYqgpb:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8873c820a01214de1176e6c45d8ee48fec8deb46/pkg/kv/dist_sender.go#L1207)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

random whitespace change
</blockquote></details>

Done.

---

*[pkg/kv/dist_sender_test.go, line 116 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/15199#-KiB0BkPJoltBCxY80qU:-KiBEjaDI6UarkTfd25H:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5832aedca608197cd9b6585c165e5fbeed223c51/pkg/kv/dist_sender_test.go#L116)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

These are test implementations. The real one is in kv/transport.
</blockquote></details>

Done.

---

*[pkg/kv/transport.go, line 163 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/15199#-KiB5d7dmWOiP_ynbKQ-:-KiBGUv95Feb2PFh3eV0:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5832aedca608197cd9b6585c165e5fbeed223c51/pkg/kv/transport.go#L163)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

The return values are backwards here.
</blockquote></details>

Done.

---

*[pkg/kv/transport.go, line 326 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/15199#-KiB6IU5NnJ5UC6A9xEq:-KiBGdPgMNaNRdyNcz2y:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5832aedca608197cd9b6585c165e5fbeed223c51/pkg/kv/transport.go#L326)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Add a TODO to plumb the lease expiration through the NotLeaseHolderError so we can target the retry for after the old lease expires. 
</blockquote></details>

Done.

---

*[pkg/sql/ambiguous_commit_test.go, line 118 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15199#-KiAuB0ZRm5pR_0ZjE7t:-KiBHidYqGftmYZzgcPB:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e83e89348f7817513ab1fe7e4a9ba8597f784407/pkg/sql/ambiguous_commit_test.go#L118)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

unnecessary condition here, just return err
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15199)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 4 of 4 files at r3.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15199)*
<!-- Sent from Reviewable.io -->
